### PR TITLE
Add an ASCII fast path for String(decoding:as:)

### PR DESCRIPTION
**Pull request by Catfish-Man**

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringCreate.swift [hunk 1]
```diff
@@ -50,18 +50,51 @@ internal func _allASCII(_ input: UnsafeBufferPointer<UInt8>) -> Bool {
...
+      )
+      repaired = true
+    }
+    
+    return (storage.asString, repaired)
+  }
+  
+  @available(*, deprecated)
+  @usableFromInline
+  internal static func _fromASCII(
```
Comment (milseman) [inline stdlib/public/core/StringCreate.swift]: Probably move to LegacyABI.swift

File: stdlib/public/core/StringCreate.swift [hunk 2]
```diff
@@ -50,18 +50,51 @@ internal func _allASCII(_ input: UnsafeBufferPointer<UInt8>) -> Bool {
...
+  
+  internal static func _fromASCIIRepairing(
+    _ input: UnsafeBufferPointer<UInt8>
+  ) -> (result: String, repairsMade: Bool) {
+    
+    if let smol = _SmallString(input) {
+      return (String(_StringGuts(smol)), false)
+    }
+    
     let storage = __StringStorage.create(initializingFrom: input, isASCII: true)
```
Comment (milseman) [inline stdlib/public/core/StringCreate.swift]: I'm surprised we don't have an invariant check for this. You can alternatively set isASCII at the end of the function.

File: stdlib/public/core/StringCreate.swift [hunk 3]
```diff
@@ -50,18 +50,51 @@ internal func _allASCII(_ input: UnsafeBufferPointer<UInt8>) -> Bool {
...
+      initializingFrom: input, isASCII: true
+    ).asString
+  }
+  
+  internal static func _fromASCIIRepairing(
+    _ input: UnsafeBufferPointer<UInt8>
+  ) -> (result: String, repairsMade: Bool) {
+    
+    if let smol = _SmallString(input) {
+      return (String(_StringGuts(smol)), false)
```
Comment (milseman) [inline stdlib/public/core/StringCreate.swift]: Where does the error correction happen here?

It seems like we need some `String(decoding:x as:ASCII.self)` test cases for invalid small and large strings.

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

**Pull request closed without merging**